src/compas/geometry/transformations/quaternions.py

```diff
@@ -0,0 +1,324 @@
+from compas.geometry.transformations import axis_and_angle_from_matrix
+from compas.geometry.transformations import matrix_from_axis_and_angle
+
+#from compas.geometry.basic import allclose
+import math
+
+
+__all__ = [
+    'quaternion_norm',
+    'quaternion_unitize',
+    'quaternion_is_unit',
+    'quaternion_multiply',
+    'quaternion_canonic',
+    'quaternion_conjugate',
+    'quaternion_from_euler_angles',
+    'euler_angles_from_quaternion',
+    'quaternion_from_axis_angle',
+    'axis_angle_from_quaternion']
+
+
+"""
+This module contains functions that operate on and/or return quaternions.
+
+Default convention to represent a quaternion in this module is a tuple of four real values [qw,qx,qy,qz].
+The first value qw is the scalar (real) part, and qx,qy,qz form the vector (complex, imaginary) part, so that:
+q = qw + qx*i + qy*j + qz*k,
+where i,j,k are basis components.
+
+Quaternion algebra basics:
+i*i = j*j = k*k = i*j*k = -1
+i*j = k    j*i = -k
+j*k = i    k*j = -i
+k*i = j    i*k = -j
+
+Quaternion as rotation:
+A rotation through an angle theta around an axis defined by a euclidean unit vector u = ux*i + uy*j + uz*k
+can be represented as a quaternion:
+q = cos(theta/2) + sin(theta/2) * [ux*i + uy*j + uz*k]
+i.e.:
+qw = cos(theta/2)
+qx = sin(theta/2) * ux
+qy = sin(theta/2) * uy
+qz = sin(theta/2) * uz
+
+For a quaternion to represent a rotation or orientation, it must be unit-length.
+
+References
+----------
+.. https://github.com/matthew-brett/transforms3d/blob/master/transforms3d/quaternions.py
+.. http://mathworld.wolfram.com/Quaternion.html
+
+"""
+
+# ----------------------------------------------------------------------
+ATOL = 1e-16  # absolute tolerance
+RTOL = 1e-3  # relative tolerance
+
+
+def isclose(a, b, atol=ATOL, rtol=RTOL):
+    # https://docs.scipy.org/doc/numpy/reference/generated/numpy.isclose.html#numpy.isclose
+    # absolute(a - b) <= (atol + rtol * absolute(b))
+
+    return abs(a - b) <= (atol + rtol * abs(b))
+
+# ----------------------------------------------------------------------
+
+
+def quaternion_norm(q):
+    """Calculates the length (euclidean norm) of a quaternion.
+
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    float
+        The length (euclidean norm) of a quaternion.
+
+    Note
+    ----
+    .. _mathworld quaternion norm
+    http://mathworld.wolfram.com/QuaternionNorm.html
+    """
+    return math.sqrt(sum([x*x for x in q]))
+
+
+def quaternion_unitize(q):
+    """Makes a quaternion unit-length.
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    list
+        A quaternion of length 1.0.
+    """
+    n = quaternion_norm(q)
+    if isclose(n, 0.0):
+        raise ValueError("The given quaternion has zero length.")
+    else:
+        return [x/n for x in q]
+
+
+def quaternion_is_unit(q):
```

Comment:
more of a point for discussion rahter than a comment.
i think functions that return a boolean should be named such that they make sense in an `if` statement.
in that sense `quaternion_is_unit` is a good name: `if quaternion_is_unit: ...`.
the rest of the framework is named differently though with the `is` at the front: `if is_quaternion_unit: ...`

what do you guys think about this?